### ISSUE TITLE
Unzip file name garbled code

### ISSUE THREAD

**Issue by l54808821** [srikanth-lingala/zip4j#69]
Unzip file name garbled code

**Comment by varunam** [comment]
I am also facing this issue. any updates on it?

**Comment by LeeYoung624** [comment]
Currently there is a workaround here [#45](https://github.com/srikanth-lingala/zip4j/issues/45)
Sample code:
```
public void test() throws ZipException, UnsupportedEncodingException {
  ZipFile zipFile = new ZipFile("/Users/someuser/Downloads/a.zip");
  for (FileHeader fileHeader : zipFile.getFileHeaders()) {
    zipFile.extractFile(fileHeader, "/Users/someuser/Downloads/extract", getEucKrEncodedFileName(fileHeader.getFileName()));
  }
}

private String getEucKrEncodedFileName(String fileName) throws UnsupportedEncodingException {
  return new String(fileName.getBytes("Cp437"), "euc-kr");
}
```

**Comment by srikanth-lingala** [comment]
@l54808821 @varunam 
It would help to provide some more information on the issue than just a one liner. With your description, it is impossible to say what is going wrong. Information like the code you used, your locale, the tool which was used to generate the zip file, and a sample zip file, etc will help to investigate an issue. Basically the more information you give, more are the chances that I will be able to help.

As @LeeYoung624 mentioned in the post above, you can try to see if that helps (replacing the charset with your local charset). This most likely is not an issue with zip4j, but it is an issue with the tool that created the zip file. There are more details in the issue mentioned by @LeeYoung624. You can look up that thread to get deeper into the issue and see if this can be related to you as well.

If you need any further help, you will have to provide more info.

**Comment by varunam** [comment]
Hi @srikanth-lingala, 

Thanks for the quick response.
I am basically trying to unzip a file which has fileName in Japanese and also the files inside zip has names in Japanese. Once unzipped, the unzipped files names are garbled. 

If I add below line of code, Japanese works fine but it breaks others.
`zf.setFileNameCharset("CP932");`

Please let me know if you need any more info. I am trying the above workaround but it doesn't seem to be working for me.

**Comment by l54808821** [comment]
zf.setFileNameCharset("CP932");
This code will solve my problem

**Comment by l54808821** [comment]
zf.setFileNameCharset("CP932");
This code will solve my problem

**Comment by l54808821** [comment]
zf.setFileNameCharset("CP932");
This code will solve my problem
This method is not available in version 2.14

**Comment by varunam** [comment]
> Hi @srikanth-lingala @LeeYoung624 ,
> 
> Thanks for the quick response.
> I am basically trying to unzip a file which has fileName in Japanese and also the files inside zip has names in Japanese. Once unzipped, the unzipped files names are garbled.
> 
> If I add below line of code, Japanese works fine but it breaks others.
> `zf.setFileNameCharset("CP932");`
> 
> Please let me know if you need any more info. I am trying the above workaround but it doesn't seem to be working for me. I have also attached the zip file for your reference.
> [ひらがなHiragara.zip](https://github.com/srikanth-lingala/zip4j/files/3652452/Hiragara.zip)

Any comments on this? @LeeYoung624 @srikanth-lingala @l54808821 ?

**Comment by LeeYoung624** [comment]
@varunam Do you mean that you got a zip that some file names are Japanese("Cp932") and some other   are another charset? And you don't know the correct charset of each file. Is it your case?
This is pretty unusual. I think using utf-8 for all these file names is a good idea.

My suggestion is to use utf-8 for these file names before being zipped. Then you can extract them with utf-8.
If you have no other choices but to unzip it, maybe you can try some charset detecting libraries with this. I don't know if there is a charset detector or not.

**Comment by varunam** [comment]
Yes, @LeeYoung624. My case is that the zip file can have names in any language and also the files inside. I don't have control over the zipping. All that I need to do is unzip uploaded file and I will probably never know how those files were zipped.

I tried utf-8 but it didn't help.

**Comment by srikanth-lingala** [comment]
@varunam @l54808821 This is not an issue with zip4j. Zip file format handles different locale filenames by encoding in utf8 if required. Unfortunately, and to my surprise, a few very popular zip tools (7zip for example) do not follow zip standards by default. 7Zip uses system charset by default for encoding file names. This really does not make sense to me and this makes compatibility between locales impossible. If someone creates a zip file in Japan (with Japanese system default charset) and send it over to a person in China, there is no way that the zip tools in China can determine that the zip file names was encoded in Japanese charset. You can override this system default charset with zip standard behaviour by an option in 7Zip preferences. But I believe most, if not all of the users, don't do this. The only supported charsets by zip format is utf8 and Cp437. Using any other charset will break the zip file compatibility.

[Here][1] is a reference thread on 7Zip forum which talks about this issue. I presume that Windows inbuilt zip utility also has the same behaviour as 7zip. The maintainer of 7zip in that thread, says that utf8 is a new feature. But that is wrong, unless in your opinion 13 years back is still "new". This feature was introduced in 2006 (The comment in the thread is from 2019). Reference: Revision 6.3.0 in [zip format specification][2].

As I said earlier, this is not an issue with zip4j, unfortunately I cannot help.

[1]: https://sourceforge.net/p/sevenzip/discussion/45797/thread/03751f07/
[2]: https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT

**Comment by srikanth-lingala** [comment]
@l54808821 
> zf.setFileNameCharset("CP932"); This code will solve my problem This method is not available in version 2.14

A similar feature will be available in the next version of zip4j. There is a PR open for this. Please follow the next release (by clicking on Watch button on top) to get notified when a new release is made.

**Comment by srikanth-lingala** [comment]
I have closed this as this is not an issue with zip4j but rather an issue that created the zip files.
